Wikipedia/Code/WMFAppViewController+Extensions.swift

```diff
@@ -97,13 +100,96 @@ extension WMFAppViewController: NotificationsCenterPresentationDelegate {
+            return
+        }
+
         let viewModel = NotificationsCenterViewModel(remoteNotificationsController: dataStore.remoteNotificationsController, languageLinkController: dataStore.languageLinkController)
         let notificationsCenterViewController = NotificationsCenterViewController(theme: theme, viewModel: viewModel)
-        navigationController?.pushViewController(notificationsCenterViewController, animated: true)
+        
+        let dismissAndPushBlock = { [weak self] in
+            self?.dismissPresentedViewControllers()
+            self?.navigationController?.pushViewController(notificationsCenterViewController, animated: true)
+        }
+
+        guard editingFlowIsInHierarchy else {
+            dismissAndPushBlock()
+            return
+        }
+        
+        presentEditorAlert(on: topMostViewController, confirmationBlock: dismissAndPushBlock)
+    }
+    
+    private var editingFlowIsInHierarchy: Bool {
+        var currentController: UIViewController? = navigationController
+
+        while let presentedViewController = currentController?.presentedViewController {
+            if let presentedNavigationController = (presentedViewController as? UINavigationController) {
+                for viewController in presentedNavigationController.viewControllers {
+                    if viewController.isPartOfEditorFlow {
+                        return true
+                    }
+                }
+            } else if presentedViewController.isPartOfEditorFlow {
+                return true
+            }
+            
+            currentController = presentedViewController
+        }
+
+        return false
+    }
+    
+    private var topMostViewController: UIViewController? {
+            
+        var topViewController: UIViewController = navigationController ?? self
+
+        while let presentedViewController = topViewController.presentedViewController {
+            topViewController = presentedViewController
+        }
+
+        return topViewController
+    }
+    
+    private func presentEditorAlert(on viewController: UIViewController, confirmationBlock: @escaping () -> Void) {
+        
+        let title = CommonStrings.editorExitConfirmationTitle
+        let message = CommonStrings.editorExitConfirmationBody
+        
+        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        let discardAction = UIAlertAction(title: CommonStrings.discardEditsActionTitle, style: .destructive) { _ in
+            confirmationBlock()
+        }
+        let cancelAction = UIAlertAction(title: CommonStrings.cancelActionTitle, style: .cancel)
+        
+        alertController.addAction(discardAction)
+        alertController.addAction(cancelAction)
+        
+        viewController.present(alertController, animated: true, completion: nil)
+        
+    }
+}
+
+fileprivate extension UIViewController {
+    var isPartOfEditorFlow: Bool {
```

Comment:
Should we also consider writing a User talk discussion reply as part of the editing flow as well? Currently, tapping a notification when replying to a User talk discussion dismisses that to show the Notifications Center.

Reply:
Great catch! I added this work to my latest commit. The awkward thing is that the talk page new topic and reply screens push on from the right, so the user actually doesn't lose their changes unless they're somehow already in a modal, say, from visiting your talk page in Settings. So I'm only displaying an alert if they are in a talk page from Settings right now (or any other modal, but I think this is the only place it can happen). Let me know what you think - it still feels a bit weird to me.